### DISCUSSION TITLE
Conditional Validation with Yup not working

### DISCUSSION THREAD

**Discussion by rohitpotato** [react-hook-form/react-hook-form#2612] [Q&A] [answered]
I have a large form that changes according to the values in one of the fields. Therefore, some fields are unregistered as a result of conditional rendering but their validation errors still occur in the `errors` object. 

I have specified a **conditional** Yup validation schema but for some reason, it's not working. React-hook-form is unable to update the validation schema accordingly. 

Here's an example:

```
securityDeposit: number()
      .required("Security Deposit amount is required.")
      .when("name", {
        is: (name) => name === "a",
        then: number().nullable().notRequired(),
        // otherwise: number().required(),
      })
```

**name** is another registered field in the schema. I have checked using the watch API and the value is being updated accordingly. 

I tried to see if the values are actually reaching the validation schema like this:

```
workStationSeats: number()
      .required("Number of Workstations are required")
      .when("name", {
        is: (name) => {
          console.log("inside validation", name);
          return name === "a";
        },
        then: number(),
        // otherwise: number().required(),
      }),
```

**Turns out name is not being logged.** 

Here's the code:

```
const methods = useForm({
    resolver: yupResolver(validationSchema),
    defaultValues: {
      workStationSeats: 0,
      collabAreaSeats: 0,
      meetingRoomSeats: 0,
      onboardingCharges: 0,
      exitCharges: 0,
      lockingPeriod: 0,
      securityDeposit: 0,
      annualEscalation: 0,
      rate: 0,
      totalDayPasses: 0,
      type: "Private Office",
    },
  });

  return <FormContext {...methods}>{children}</FormContext>;
```

**I was using version 6.0.6 and recently updated to 6.4.0 but the problem still persists. Any help is appreciated!!!**

**Comment by bluebill1049** [comment]
can you share a simple codesandbox for this issue?

**Reply by rohitpotato** [reply]
Sure: https://codesandbox.io/s/youthful-fog-4lw31?file=/src/FormContext.js

Steps to reproduce:

1. By default checkbox is false, therefore the field `workStationSeats` is not displayed which means it is not registered.

2. Click the checkbox to enable the `workStationSeats` field.

3. Even after specifying in the rules in the Yup validation schema that `workStationSeats` can be treated as `nullable` and `notRequired` when **checkbox is false**, it triggers a validation error. (Click on submit without clicking the checkbox, it triggers an error  for "workStationSeats" with the conditional validation in place.

Also, I have tried many combinations of **is, then, and otherwise** but nothing seems to work.

Thank you for your time!!!!

**Reply by rohitpotato** [reply]
> can you share a simple codesandbox for this issue?

Hey @bluebill1049 , I have shared a codesandbox for the issue? It would be great if you could look into it.
Thanks!

**Reply by bluebill1049** [reply]
hey just took a quick look at the codesandbox, I see the problem it's because dependency validation, you will have to use `trigger` to trigger dependant validation. we focus single input at a time.

**Comment by rohitpotato** [comment, marked as answer]
Do you mean monitoring the value of the field using the watch API and useEffect and using the trigger method to trigger validation? I tried that and I am still facing the same issue

**Reply by bluebill1049** [reply]
I meant attach onChange to trigger the dependent validation. worth to `trigger()` the whole form to see the schema is working first as well

**Reply by rohitpotato** [reply]
> I meant attach onChange to trigger the dependent validation. worth to `trigger()` the whole form to see the schema is working first as well

Sorry but I think I am unable to understand correctly. `onChange` handler on what exactly? If the field is unregistered because of conditional rendering, the `onChange` will have no effect, right? And if you are talking `onChange` for the `checkbox` field, I don't understand how exactly that fits into the picture.

**Reply by bluebill1049** [reply]
`<input type="checkbox" name="isActive" onChange={() => trigger()} ref={register} />` is what i meant.

**Reply by rohitpotato** [reply]
Didn't work.

**Reply by bluebill1049** [reply]
if that still not working, i would recommend build a custom resolver and manually test out your schema. https://codesandbox.io/s/react-hook-form-customresoliver-v6-h4ti6

**Reply by bluebill1049** [reply]
there is no magic with how react hook form validate with yup schema, we supply the data and run against the schema, that's all.

**Reply by bluebill1049** [reply]
also RHF only returns `string` for native input. in case, that's the root cause of the issue.

**Reply by rohitpotato** [reply]
Not using Boolean also yields the same result. I have tried it with a string-based input as well but with no avail. Can this be a potential bug?

**Reply by bluebill1049** [reply]
have you tried the custom resolver? the one i suggested above, run through form value against your schema.

**Reply by rohitpotato** [reply]
Yes, I did, I plugged it in with the `useForm` configuration, but it's not working. It's throwing some errors which were enough to understand that the problem still exists.

**Reply by rohitpotato** [reply]
@bluebill1049 I feel like I am bugging you but it would really awesome to know where exactly the issue is. Is it a bug, or the expected behaviour and if there's something wrong with the way I am doing this? Any insight is appreciated.

**Reply by bluebill1049** [reply]
no man, It was 12 last night and it's getting late. I will do a test in code today (i have bookmarked your question). I still believe it's schema related, let's me check.

**Reply by rohitpotato** [reply]
Great!!

**Reply by bluebill1049** [reply]
<img width="951" alt="Screen Shot 2020-08-19 at 9 27 48 am" src="https://user-images.githubusercontent.com/10513364/90575208-4263c900-e1fe-11ea-84ad-bd6735cb8d0e.png">
please take a look above, it's your schema which causing this issue.

The schema throw an error because `workStationSeats` doesn't exist. just like i suggested take a look at the codesandbox: you will see what data RHF is passing through and why it's broken. there is no magic.

**Reply by bluebill1049** [reply]
https://codesandbox.io/s/competent-wave-h1ogd?file=/src/FormContext.js

**Reply by rohitpotato** [reply]
Yes I figured out it out. Turns out if I reverse the conditions it works fine. That is, setting default behaviour to not required and making it required once `isActive` is true. 

No idea why but I am glad it works. (I'll investigate this)

 Thank you for your time bill!! Your library is great and I use it for every single form in my project.

**Reply by bluebill1049** [reply]
np :)
